### PR TITLE
Initialize the AuditEvent with the AuditContext

### PR DESCRIPTION
**Pull request by tallclair**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Always initialize the audit event at the beginning of the request chain, and store audit data directly on the event.

#### Which issue(s) this PR fixes:
For https://github.com/kubernetes/kubernetes/issues/109087

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/sig auth api-machinery
/milestone v1.26

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/audit/context.go
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/audit/context.go:146]: dropping the metalevel check is intentional?

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/audit/context.go:146]: I guess it is, and replaced through the check in  `Enabled()`

**Comment by sttts** [comment]
Lgtm.

As this is pretty deep in the handler chain, maybe another set of eyes would be helpful. @tkashem @benluddy does one of you have a moment to double check?

**Comment by tkashem** [comment]
> Lgtm.
> 
> As this is pretty deep in the handler chain, maybe another set of eyes would be helpful. @tkashem @benluddy does one of you have a moment to double check?

Yes, I will take a look at it next week, thanks!
/assign

File: staging/src/k8s.io/apiserver/pkg/audit/context.go
Comment (ibihim) [inline staging/src/k8s.io/apiserver/pkg/audit/context.go:164]: Nit: Is it always initialized with an ID, so that we don't need to check for empty string as it initialized by `WithAuditInit`. Isn't it making it easier to make a mistake if the audit pkg is used outside of `k8s.io/apiserver`? I could imagine that it takes some debugging to figure out that it can be true on an empty string.

File: staging/src/k8s.io/apiserver/pkg/audit/request.go
Comment (ibihim) [inline staging/src/k8s.io/apiserver/pkg/audit/request.go:46]: We use the audit package for or [openshift/oauth-server](https://github.com/openshift/oauth-server) and the removal of `NewEventFromRequest` would break our tests. So I rewrote them against your PR and the usage of the audit package is now really nice. All the changes around the removal of `*[]annotations` is amazing! 🎉

File: staging/src/k8s.io/apiserver/pkg/audit/context.go
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/audit/context.go:43]: was probably just a clever way to distinguish between enabled and disabled auditing

Comment (tallclair) [reply staging/src/k8s.io/apiserver/pkg/audit/context.go:43]: We used to use a pointer since the event gets initialized later in the handler stage, so it was nil for the first part. That's why we needed the separate annotations slice so earlier handlers (e.g. authn) could record annotations.

This change switches away from a pointer so we can start accumulating information (annotations) on the event right away. Sure, we could keep it a pointer and just populate an empty event right away, but I don't see any reason for that.

File: staging/src/k8s.io/apiserver/pkg/audit/context.go
Comment (tallclair) [reply staging/src/k8s.io/apiserver/pkg/audit/context.go:164]: Yes, the audit ID is always initialized at the same time as the audit context: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit_init.go#L41-L50

But you bring up a good point. Looking at everywhere this is used, I think the bool return value is really meant to indicate whether auditing is enabled. This method isn't used in very many places, so I think we should probably just inline it with an explicit `auditContext.Enabled()` check instead.

Comment (tallclair) [reply staging/src/k8s.io/apiserver/pkg/audit/context.go:164]: Actually, since the eventual goal is to make the `AuditContext.Event` non-exported, I think it makes sense to keep this function as-is for now.

**Comment by ibihim** [comment]
lgtm

File: staging/src/k8s.io/apiserver/pkg/audit/context_test.go
Comment (tkashem) [inline staging/src/k8s.io/apiserver/pkg/audit/context_test.go:96]: instead of removing the test, can we rewrite it using `AddAuditAnnotation`?
```
func TestLogAnnotation(t *testing.T) {
        ctx := WithAuditContext(context.TODO())
        ev := AuditEventFrom(ctx)
	ev.Level := auditinternal.LevelMetadata
        ev.AuditID = "fake id"

	AddAuditAnnotation(ctx, "foo", "bar")
	AddAuditAnnotation(ctx, "foo", "baz")
	assert.Equal(t, "bar", ev.Annotations["foo"], "audit annotation should not be overwritten.")

	AddAuditAnnotation(ctx, "qux", "")
	AddAuditAnnotation(ctx, "qux", "baz")
	assert.Equal(t, "", ev.Annotations["qux"], "audit annotation should not be overwritten.")
}
```

File: staging/src/k8s.io/apiserver/pkg/audit/context.go
Comment (tkashem) [inline staging/src/k8s.io/apiserver/pkg/audit/context.go:51]: I think the initial default should return `false`, right?
```
func TestEnabled(t *testing.T) {
	ac := &AuditContext{}
	if ac.Enabled() {
		t.Errorf("expected the default to return false")
	}
}
```
maybe add `len(ac.Event.Level) > 0`

File: staging/src/k8s.io/apiserver/pkg/audit/context.go
Comment (tkashem) [inline staging/src/k8s.io/apiserver/pkg/audit/context.go:131]: I think it's safer for the the initial context to return a `LevelNone` for the `Level` field

```
func WithAuditContext(parent context.Context) context.Context {
	if AuditContextFrom(parent) != nil {
		return parent // Avoid double registering.
	}
        ac := &AuditContext{
              Event: {
                  Level: LevelNone
              }
        }
	return genericapirequest.WithValue(parent, auditKey, ac)
}
```

https://github.com/kubernetes/kubernetes/blob/17bf864c1fc20e4badc9a6e0659c5190310462e1/staging/src/k8s.io/apiserver/pkg/audit/context.go#L159-L166

File: staging/src/k8s.io/apiserver/pkg/server/config_test.go
Comment (tkashem) [inline staging/src/k8s.io/apiserver/pkg/server/config_test.go:286]: maybe we should confirm here that the level is set to `LevelNone`?

**Review by tkashem — COMMENTED** [review]
this LGTM, I have a few minor comments, thanks!

File: staging/src/k8s.io/apiserver/pkg/audit/context.go
Comment (tkashem) [inline staging/src/k8s.io/apiserver/pkg/audit/context.go:155]: I think storing the `auditID` in the request context is orthogonal to audit logging. A cluster operator can disable audit logging, but we would still want to store the auditID stored in the request context. The auditID is also propagated to the aggregated APIServer(s). 

We have some logs that emit the auditID, for example: 
https://github.com/kubernetes/kubernetes/blob/4db6bde859d912e11ab081a06e8ea17b2e044f5d/staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go#L58
I think we would want to keep the correlation here, even though audit logging is disabled, thoughts?

File: staging/src/k8s.io/apiserver/pkg/audit/context.go
Comment (tkashem) [inline staging/src/k8s.io/apiserver/pkg/audit/context.go:63]: I don't think `ac == nil` can be replaced with `ac.Enabled`. If the handler chain is setup properly we should never see `ac` to be `nil`, irrespective of whether audit logging is disabled. 

`// auditing is not enabled` is probably misleading, I apologize for that.  it's more like, `we should never be here if the handler chain is setup properly`

`WithAudit` filter is where we decide whether audit logging is enabled/disabled, but some filters that precede `WithAudit` filter need to add some annotations, esp. during authentication, for example:

https://github.com/kubernetes/kubernetes/blob/4db6bde859d912e11ab081a06e8ea17b2e044f5d/staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go#L136

These annotations are added before we decide whether audit logging is enabled. `ac.Enabled` is going to return `false` before that decision is made, and we are going to drop these annotations which is not expected behavior I believe.

File: staging/src/k8s.io/apiserver/pkg/audit/context.go
Comment (tkashem) [inline staging/src/k8s.io/apiserver/pkg/audit/context.go:78]: same as the above comment

File: staging/src/k8s.io/apiserver/pkg/audit/context.go
Comment (tkashem) [inline staging/src/k8s.io/apiserver/pkg/audit/context.go:97]: same as the other comment

File: staging/src/k8s.io/apiserver/pkg/audit/request.go
Comment (tallclair) [reply staging/src/k8s.io/apiserver/pkg/audit/request.go:59]: No, I think it's safe to remove. Done.

File: staging/src/k8s.io/apiserver/pkg/audit/context_test.go
Comment (tallclair) [reply staging/src/k8s.io/apiserver/pkg/audit/context_test.go:96]: There's already a `TestAddAuditAnnotation`, so I just added a case to cover overlapping annotations.

File: staging/src/k8s.io/apiserver/pkg/audit/context.go
Comment (tallclair) [reply staging/src/k8s.io/apiserver/pkg/audit/context.go:51]: No, this is intentional. If annotations (or any other request metadata) are recorded before the policy is evaluated, we want to store them. Once the policy is evaluated to `None` we can stop doing the work in that case. I'll add a comment to this affect.

File: staging/src/k8s.io/apiserver/pkg/audit/context.go
Comment (tallclair) [reply staging/src/k8s.io/apiserver/pkg/audit/context.go:131]: See above comment about capturing data before the policy is evaluated.

File: staging/src/k8s.io/apiserver/pkg/server/config_test.go
Comment (tallclair) [reply staging/src/k8s.io/apiserver/pkg/server/config_test.go:286]: (See above comments)

File: staging/src/k8s.io/apiserver/pkg/audit/context.go
Comment (tallclair) [reply staging/src/k8s.io/apiserver/pkg/audit/context.go:155]: Yeah, that makes sense. You might also be interested in this issue, which has come up a couple of times in sig-apimachinery: https://github.com/kubernetes/kubernetes/issues/101597

File: staging/src/k8s.io/apiserver/pkg/audit/context.go
Comment (tallclair) [reply staging/src/k8s.io/apiserver/pkg/audit/context.go:63]: Ah, yeah, this ties in to the above discussion about how `Enabled()` should behave before the policy is evaluated. The approach I took is to treat it as enabled before the policy is evaluated, for exactly the reason you describe here. The other approach is to default to disabled, and use a `nil`-check here.

The advantage to my approach is once we have evaluated the policy, and explicitly decided to disabled auditing (level None), then we can skip some work here. In practice, the difference is probably negligible.

WDYT?

**Comment by tallclair** [comment]
Thanks for the reviews. Feedback addressed, PTAL.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go
Comment (tkashem) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go:126]: with the current approach `ac.Enabled` should always return true here, it could also be a little confusing to the reader - the function that is evaluating audit policy is checking whether policy has been evaluated already, 
we could remove the check entirely, or bring back the nil check here, thoughts?

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authn_audit.go
Comment (tkashem) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/authn_audit.go:46]: It seems the `Enabled` function is used outside of its package where the caller does not not have the pre-evaluation constraints we have inside the `audit` package, so for future maintainability i think the following might be useful:
```
func (ac *AuditContext) Enabled() bool {
	return ac != nil && len(ac.RequestAuditConfig.Level) > 0 && ac.RequestAuditConfig.Level != auditinternal.LevelNone
}
```

inside the `audit` package, where we have to be mindful of the pre-evaluation constraint, we can use the following private function:
```
func (ac *AuditContext) shouldAddAnnotation() bool {
	switch {
	case ac == nil:
		return false
	case ac.RequestAuditConfig.Level == "":
		return true
	}
	return ac.Enabled()
}
```

```
func AddAuditAnnotation(ctx context.Context, key, value string) {
	ac := AuditContextFrom(ctx)
	if !ac.shouldAddAnnotation() {
		return
	}
...
```

- this approach does not leak the constraint outside of the `audit` package, the invoker of `Enabled` is aware that policy has been evaluated
- `AddAnnotation*` functions have the pre-evaulation constraint and these functions can resort to an internal `shouldAddAnnotation`
Thoughts?

File: staging/src/k8s.io/apiserver/pkg/audit/context_test.go
Comment (tallclair) [reply staging/src/k8s.io/apiserver/pkg/audit/context_test.go:85]: Done.

File: staging/src/k8s.io/apiserver/pkg/audit/context_test.go
Comment (tallclair) [reply staging/src/k8s.io/apiserver/pkg/audit/context_test.go:96]: Thanks, added.

File: staging/src/k8s.io/apiserver/pkg/audit/context.go
Comment (tallclair) [reply staging/src/k8s.io/apiserver/pkg/audit/context.go:50]: Personally I prefer the shorter name, since this is on the `AuditContext`, I don't think the extra context on the name is needed. I'm open to making the change if you feel strongly though.

File: staging/src/k8s.io/apiserver/pkg/audit/context.go
Comment (tallclair) [reply staging/src/k8s.io/apiserver/pkg/audit/context.go:65]: I think that test case is already covered by `TestEnabled`

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go
Comment (tallclair) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go]: Yeah, that makes sense. Done.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authn_audit.go
Comment (tallclair) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/authn_audit.go:46]: I'm afraid I don't entirely follow this comment.

In the next audit refactor pass, I'd like to make `ac.Event` non-exported, and enforce that all event handling be done within the audit package, which I think will make this comment moot.

If you're OK with it, I propose we leave this as-is for now (the logic matches the previous logic in practice, as the event initialization & policy evaluation happen in the same step), and revisit this in the follow-up PR. WDYT?

**Review by tallclair — COMMENTED** [review]
Sorry for the delay in getting back to this. Responded on a few of your comments. PTAL.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authn_audit.go
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/authn_audit.go:46]: yes, that sounds good to me.

File: staging/src/k8s.io/apiserver/pkg/audit/context.go
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/audit/context.go:50]: sounds good, i think when we make the `Event` object non-exported, it will become moot as well.

**Comment by tkashem** [comment]
@tallclair this LGTM from my side, are you going to squash the fix-ups?

**Comment by tallclair** [comment]
The label `tide/merge-method-squash` means that the extra commits get squashed on merge.

**Comment by tallclair** [comment]
@tkashem thank you for your reviews!

**Comment by tkashem** [comment]
/lgtm

Thanks!

**Comment by liggitt** [comment]
/unassign
unfortunately I don't have bandwidth to pick up this review before freeze

**Comment by tallclair** [comment]
/milestone v1.28

Punting this to the next release. Hopefully we can get it merged early in the cycle.

File: staging/src/k8s.io/apiserver/pkg/audit/context.go
Comment (stlaz) [reply staging/src/k8s.io/apiserver/pkg/audit/context.go:165]: According to the updated description, `ac.Enabled()` should be returned instead, in case the policy level is "None".
I am not sure whether that would still align with the original purpose of the bool.

**Comment by furkatgofurov7** [comment]
Hey! v1.28 Bug Triage Lead here.

I am reaching out to know if this PR is on track for the 1.28 release? Thanks in advance!

File: staging/src/k8s.io/apiserver/pkg/audit/context.go
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/audit/context.go:165]: Is this comment still open?

**Comment by furkatgofurov7** [comment]
@tallclair hey! Can we please label this with priority? 

Also, just a reminder that the code freeze is starting [01:00 UTC Wednesday 19th July 2023 / 18:00 PDT Tuesday 18th July 2023](https://everytimezone.com/s/72ee8496) (about 3 weeks from now) and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.

File: staging/src/k8s.io/apiserver/pkg/audit/context.go
Comment (stlaz) [inline staging/src/k8s.io/apiserver/pkg/audit/context.go:62]: This godoc mentions functions that no longer exist, please fix that. Perhaps the changes from https://github.com/kubernetes/kubernetes/pull/117406 might be applicable here?

**Review by stlaz — CHANGES_REQUESTED** [review]
/lgtm cancel
https://github.com/kubernetes/kubernetes/pull/113611/files#r1231007227 has not been addressed yet

**Comment by stlaz** [comment]
/lgtm
/hold cancel
the comments are non-blocking, can be addressed as follow-up

**Comment by tallclair** [comment]
Thanks for getting this merged. I'm on leave right now, but I'll make sure to pick this back up in the next release cycle.

**Pull request merged**
